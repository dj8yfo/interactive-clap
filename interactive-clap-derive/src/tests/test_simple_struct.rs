fn pretty_codegen(ts: &proc_macro2::TokenStream) -> String {
    let file = syn::parse_file(&ts.to_string()).unwrap();
    prettyplease::unparse(&file)
}

#[test]
fn test_simple_struct() {
    let input = syn::parse_quote! {
        struct Args {
            age: u64,
            first_name: String,
            second_name: String,
        }
    };

    let interactive_clap_codegen = crate::derives::interactive_clap::impl_interactive_clap(&input);
    insta::assert_snapshot!(pretty_codegen(&interactive_clap_codegen));

    let to_cli_args_codegen = crate::derives::to_cli_args::impl_to_cli_args(&input);
    insta::assert_snapshot!(pretty_codegen(&to_cli_args_codegen));
}

#[test]
fn test_flag() {
    let input = syn::parse_quote! {
        struct Args {
            /// Offline mode
            #[interactive_clap(long)]
            offline: bool
        }
    };

    let interactive_clap_codegen = crate::derives::interactive_clap::impl_interactive_clap(&input);
    insta::assert_snapshot!(pretty_codegen(&interactive_clap_codegen));

    let input = syn::parse_quote! {
        struct CliArgs {
            /// Offline mode
            #[clap(long)]
            offline: bool
        }
    };

    let to_cli_args_codegen = crate::derives::to_cli_args::impl_to_cli_args(&input);
    insta::assert_snapshot!(pretty_codegen(&to_cli_args_codegen));
}

#[test]
fn test_vec_multiple_opt() {
    let input = syn::parse_quote! {
        struct Args {
            #[interactive_clap(long)]
            #[interactive_clap(vec_multiple_opt)]
            pub env: Vec<String>,
        }
    };

    let interactive_clap_codegen = crate::derives::interactive_clap::impl_interactive_clap(&input);
    insta::assert_snapshot!(pretty_codegen(&interactive_clap_codegen));

}

#[test]
fn test_vec_multiple_opt_to_cli_args() {
    let input = syn::parse_quote! {
        pub struct CliArgs {
            #[clap(long)]
            pub env: Vec<String>,
        }
    };

    let to_cli_args_codegen = crate::derives::to_cli_args::impl_to_cli_args(&input);
    insta::assert_snapshot!(pretty_codegen(&to_cli_args_codegen));
}
