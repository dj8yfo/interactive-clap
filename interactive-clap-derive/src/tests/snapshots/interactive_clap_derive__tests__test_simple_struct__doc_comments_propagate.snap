---
source: interactive-clap-derive/src/tests/test_simple_struct.rs
expression: pretty_codegen(&interactive_clap_codegen)
---
#[derive(Debug, Default, Clone, clap::Parser, interactive_clap::ToCliArgs)]
#[clap(author, version, about, long_about = None)]
pub struct CliArgs {
    /// short first field description
    ///
    /// a longer paragraph, describing the usage and stuff with first field's
    /// awarenes of its possible applications
    pub first_field: Option<<u64 as interactive_clap::ToCli>::CliVariant>,
    /// short second field description
    ///
    /// a longer paragraph, describing the usage and stuff with second field's
    /// awareness of its possible applications
    pub second_field: Option<<String as interactive_clap::ToCli>::CliVariant>,
}
impl interactive_clap::ToCli for Args {
    type CliVariant = CliArgs;
}
pub struct InteractiveClapContextScopeForArgs {
    pub first_field: u64,
    pub second_field: String,
}
impl interactive_clap::ToInteractiveClapContextScope for Args {
    type InteractiveClapContextScope = InteractiveClapContextScopeForArgs;
}
impl interactive_clap::FromCli for Args {
    type FromCliContext = ();
    type FromCliError = color_eyre::eyre::Error;
    fn from_cli(
        optional_clap_variant: Option<<Self as interactive_clap::ToCli>::CliVariant>,
        context: Self::FromCliContext,
    ) -> interactive_clap::ResultFromCli<
        <Self as interactive_clap::ToCli>::CliVariant,
        Self::FromCliError,
    >
    where
        Self: Sized + interactive_clap::ToCli,
    {
        let mut clap_variant = optional_clap_variant.clone().unwrap_or_default();
        if clap_variant.first_field.is_none() {
            clap_variant
                .first_field = match Self::input_first_field(&context) {
                Ok(Some(first_field)) => Some(first_field),
                Ok(None) => {
                    return interactive_clap::ResultFromCli::Cancel(Some(clap_variant));
                }
                Err(err) => {
                    return interactive_clap::ResultFromCli::Err(Some(clap_variant), err);
                }
            };
        }
        let first_field = clap_variant.first_field.clone().expect("Unexpected error");
        if clap_variant.second_field.is_none() {
            clap_variant
                .second_field = match Self::input_second_field(&context) {
                Ok(Some(second_field)) => Some(second_field),
                Ok(None) => {
                    return interactive_clap::ResultFromCli::Cancel(Some(clap_variant));
                }
                Err(err) => {
                    return interactive_clap::ResultFromCli::Err(Some(clap_variant), err);
                }
            };
        }
        let second_field = clap_variant.second_field.clone().expect("Unexpected error");
        let new_context_scope = InteractiveClapContextScopeForArgs {
            first_field: first_field.into(),
            second_field: second_field.into(),
        };
        interactive_clap::ResultFromCli::Ok(clap_variant)
    }
}
impl Args {
    fn input_first_field(_context: &()) -> color_eyre::eyre::Result<Option<u64>> {
        match inquire::CustomType::new(
                concat!(
                    r" short first field description", r"",
                    r" a longer paragraph, describing the usage and stuff with first field's",
                    r" awarenes of its possible applications",
                )
                    .trim(),
            )
            .prompt()
        {
            Ok(value) => Ok(Some(value)),
            Err(
                inquire::error::InquireError::OperationCanceled
                | inquire::error::InquireError::OperationInterrupted,
            ) => Ok(None),
            Err(err) => Err(err.into()),
        }
    }
    fn input_second_field(_context: &()) -> color_eyre::eyre::Result<Option<String>> {
        match inquire::CustomType::new(
                concat!(
                    r" short second field description", r"",
                    r" a longer paragraph, describing the usage and stuff with second field's",
                    r" awareness of its possible applications",
                )
                    .trim(),
            )
            .prompt()
        {
            Ok(value) => Ok(Some(value)),
            Err(
                inquire::error::InquireError::OperationCanceled
                | inquire::error::InquireError::OperationInterrupted,
            ) => Ok(None),
            Err(err) => Err(err.into()),
        }
    }
    pub fn try_parse() -> Result<CliArgs, clap::Error> {
        <CliArgs as clap::Parser>::try_parse()
    }
    pub fn parse() -> CliArgs {
        <CliArgs as clap::Parser>::parse()
    }
    pub fn try_parse_from<I, T>(itr: I) -> Result<CliArgs, clap::Error>
    where
        I: ::std::iter::IntoIterator<Item = T>,
        T: ::std::convert::Into<::std::ffi::OsString> + ::std::clone::Clone,
    {
        <CliArgs as clap::Parser>::try_parse_from(itr)
    }
}
impl From<Args> for CliArgs {
    fn from(args: Args) -> Self {
        Self {
            first_field: Some(args.first_field.into()),
            second_field: Some(args.second_field.into()),
        }
    }
}
